// fortmatic/webpack.config.js
var path = require('path');

var configNPM = {
    mode: 'production',
    entry: ['babel-polyfill', './src/index.js'],
    output: {
        path: path.resolve(__dirname, "lib"),
        filename: "fortmatic.js",
        libraryTarget: 'commonjs2',
        // library: 'Fortmatic'
    },
    module: {
        rules: [
            {
                test: /\.js$/,
                exclude: /(node_modules)/,
                use: {
                    loader: "babel-loader"
                }
            }
        ]
    }
};

var configCDN = {
    mode: 'production',
    entry: ['babel-polyfill','./src/index.js'],
    output: {
        path: path.resolve(__dirname, "dist"),
        filename: "fortmatic.js",
        libraryTarget: 'window',
        library: 'Fortmatic'
    },
    module: {
        rules: [
            {
                test: /\.js$/,
                exclude: /(node_modules)/,
                use: {
                    loader: "babel-loader"
                }
            }
        ]
    }
};

module.exports = [configNPM, configCDN];
